src/train.js

```diff
@@ -141,50 +145,59 @@ async function trainOnTabs() {
-    await setViewportSize(tabs[0], 1024, 768);  // for consistent element sizing in samples due to text wrap, etc.
+    // Show progress bar and output.
+    g_progressBar.classList.remove('hidden');
+    document.getElementById('output').classList.remove('hidden');
+
+    try {
 
-    const rulesetName = document.getElementById('ruleset').value;
-    const tuner = new Tuner(tabs, rulesetName);
-    const [tunedCoeffs, tunedCost] = await tuner.anneal(updateProgress);
+        // TODO: Using "active" here rather than a tab ID presents a race condition
+        // if you quickly switch away from the tab after clicking the Train button.
+        const tabs = (await browser.tabs.query({currentWindow: true, active: false}));
+        await setViewportSize(tabs[0], 1024, 768);  // for consistent element sizing in samples due to text wrap, etc.
 
-    progressBar.setAttribute('style', 'display: none');
-    progressBar.setAttribute('value', 0);
-    trainButton.disabled = false;
+        const rulesetName = document.getElementById('ruleset').value;
+        const tuner = new Tuner(tabs, rulesetName);
+        await tuner.anneal(updateProgress);
+
+    } finally {
+        // Restore UI state, leaving output visible.
+        g_progressBar.classList.add('hidden');
+        g_progressBar.setAttribute('value', 0);
+        trainButton.disabled = false;
+    }
 }
 
 function empty(element) {
-    while (element.firstChild) {
-        element.removeChild(element.firstChild);
+    while (element.lastChild) {
+        element.removeChild(element.lastChild);
     }
 }
 
 function updateProgress(ratio, bestSolution, bestCost, successesOrFailures) {
-    document.getElementById('progress').setAttribute('value', ratio);
-
-    // Update best coeffs and accuracy:
-    const coeffsDiv = document.getElementById('coeffs');
-    const accuracyDiv = document.getElementById('accuracy');
-    Array.prototype.forEach.call(document.getElementsByClassName('output'),
-                                 e => e.classList.remove('output'));
-    coeffsDiv.classList.remove('output');
-    empty(coeffsDiv);
-    empty(accuracyDiv);
-    coeffsDiv.appendChild(document.createTextNode(`[${bestSolution}]`));
-    accuracyDiv.appendChild(document.createTextNode(`${((1 - bestCost) * 100).toFixed(1)}%`));
+    // Tick progress meter.
+    g_progressBar.setAttribute('value', ratio);
+
+    // Update best coeffs and accuracy.
+    g_coeffs_div.firstChild.textContent = `[${bestSolution}]`;
+    g_accuracy_div.firstChild.textContent = `${((1 - bestCost) * 100).toFixed(1)}%`;
+
     if (successesOrFailures) {
-        const goodBadDiv = document.getElementById('goodBad');
-        empty(goodBadDiv);
+        if (g_goodBadDiv.childElementCount !== successesOrFailures.length) {
+            empty(g_goodBadDiv);
+            for (let _ of successesOrFailures) {
+                let div = document.createElement('div');
```

Comment:
Again with the `const`. You're just doing this to prod me to get linting going. ;-)